x/programs/simulator/ffi/ffi.go

```diff
@@ -0,0 +1,155 @@
+
+	rt := runtime.NewRuntime(runtime.NewConfig(), SimLogger)
+	result, err := rt.CallProgram(SimContext, callInfo)
+	if err != nil {
+		return newCallProgramResponse(nil, 0, fmt.Errorf("error during runtime execution: %w", ErrRuntimeExecution))
+	}
+
+	fuel := callInfo.RemainingFuel()
+	return newCallProgramResponse(result, fuel, nil)
+}
+
+func createRuntimeCallInfo(db state.Mutable, ctx *C.SimulatorCallContext) *runtime.CallInfo {
+	paramBytes := C.GoBytes(unsafe.Pointer(ctx.params.data), C.int(ctx.params.length))
+	methodName := C.GoString(ctx.method)
+	actorBytes := C.GoBytes(unsafe.Pointer(&ctx.actor_address), codec.AddressLen)
+	programBytes := C.GoBytes(unsafe.Pointer(&ctx.program_address), codec.AddressLen)
+
+	return &runtime.CallInfo{
+		State:        simState.NewProgramStateManager(db),
+		Actor:        codec.Address(actorBytes),
+		FunctionName: methodName,
+		Program:      codec.Address(programBytes),
+		Params:       paramBytes,
+		Fuel:         uint64(ctx.max_gas),
+		Height:       uint64(ctx.height),
+		Timestamp:    uint64(ctx.timestamp),
+	}
+}
+
+//export CreateProgram
+func CreateProgram(db *C.Mutable, path *C.char) C.CreateProgramResponse {
+	state := simState.NewSimulatorState(unsafe.Pointer(db))
+	programManager := simState.NewProgramStateManager(state)
+
+	programPath := C.GoString(path)
+	programBytes, err := os.ReadFile(programPath)
+	if err != nil {
+		return C.CreateProgramResponse{
+			error: C.CString(err.Error()),
+		}
+	}
+
+	programID, err := generateRandomID()
+	if err != nil {
+		return C.CreateProgramResponse{
+			error: C.CString(err.Error()),
+		}
+	}
+
+	err = programManager.SetProgram(context.TODO(), programID, programBytes)
+	if err != nil {
+		errmsg := "program creation failed: " + err.Error()
+		return C.CreateProgramResponse{
+			error: C.CString(errmsg),
+		}
+	}
+
+	account, err := programManager.DeployProgram(context.TODO(), programID, []byte{})
+	if err != nil {
+		errmsg := "program deployment failed: " + err.Error()
+		return C.CreateProgramResponse{
+			error: C.CString(errmsg),
+		}
+	}
+	return C.CreateProgramResponse{
+		error: nil,
+		program_id: C.ID{
+			*(*[32]C.uchar)(C.CBytes(programID[:])),
+		},
+		program_address: C.Address{
+			*(*[33]C.uchar)(C.CBytes(account[:])),
+		},
+	}
+}
+
+// generateRandomID creates a unique ID.
+// Note: ids.GenerateID() is not used because the IDs are not unique and will
+// collide.
+func generateRandomID() (ids.ID, error) {
+	key := make([]byte, 32)
+	_, err := rand.Read(key)
+	if err != nil {
+		return ids.Empty, err
+	}
+	id, err := ids.ToID(key)
+	if err != nil {
+		return ids.Empty, err
+	}
+
+	return id, nil
+}
+
+func newCallProgramResponse(result []byte, fuel uint64, err error) C.CallProgramResponse {
+	var errPtr *C.char
+	if err == nil {
+		errPtr = nil
+	} else {
+		errPtr = C.CString(err.Error())
+	}
+
+	return C.CallProgramResponse{
+		error: errPtr,
+		result: C.Bytes{
+			data:   (*C.uint8_t)(C.CBytes(result)),
```

Comment:
`C.CBytes` mallocs so we need to call `C.Free`. Should we expose an external `free` method in go so that rust can perform this cleanup?